cmflib/contrib/analysis_engine.py

```diff
@@ -0,0 +1,263 @@
+###
+# Copyright (2023) Hewlett Packard Enterprise Development LP
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###
+
+"""
+This module defines a number of analysis engines that provide domain- or task-specific analytic functions that can
+be performed on AI pipeline metadata.
+
+# Utility classes
+Different analysis engines can consume data in various flavours, such as tabular- or graph-based representation. The
+`MetricSource` class can be used to enable some or all of the implemented analysis engines access data stored
+differently. Internally, two classes  are implemented -  `_TabularMetricSource` and `_GraphMetricSource`.
+
+
+# Analysis engine classes
+Each analysis engine implements one or multiple related analytic functions. The following analysis engines are
+implemented.
+
+## Parameter sensitivity analysis engine
+This engine can be used to identify how a machine learning metric like accuracy depends on a particular parameter such
+as learning rate.
+"""
+
+import abc
+import typing as t
+
+import numpy as np
+import pandas as pd
+from contrib.graph_api import Artifact, Type
+
+__all__ = ["MetricSource", "ParameterSensitivityReport", "ParameterSensitivityAnalysisEngine"]
+
+
+class MetricSource(abc.ABC):
+    """Class that provides standard interface for getting metric values.
+
+    This class is public to enable users define their own sources.
+    """
+
+    @abc.abstractmethod
+    def values(self, metric_name: str) -> np.ndarray:
+        """Retrieve metric values for the given metric.
+
+        Args:
+            metric_name: Name of a metric.
+        Returns:
+            One-rank numpy array containing metric values.
+        """
+        raise NotImplementedError
+
+
+class ParameterSensitivityReport:
```

Comment:
@sergey-serebryakov should'nt we consider the parameter values also for sensitivity report ?